aws/aws.go

```diff
@@ -211,11 +226,83 @@ func (s *ec2Ops) waitAttachmentStatus(
 		fmt.Sprintf("Invalid volume object for volume %s", volumeID), "")
 }
 
-func (s *ec2Ops) Name() string { return "aws" }
+func (s *awsOps) Name() string { return "aws" }
+
+func (s *awsOps) InstanceID() string { return s.instance }
+
+func (s *awsOps) InspectInstance(instanceID string) (*cloudops.InstanceInfo, error) {
+	inst, err := DescribeInstanceByID(s.ec2, instanceID)
+	if err != nil {
+		return nil, err
+	}
 
-func (s *ec2Ops) InstanceID() string { return s.instance }
+	instInfo := &cloudops.InstanceInfo{
+		CloudResourceInfo: cloudops.CloudResourceInfo{
+			Name:   instanceID,
+			ID:     *inst.InstanceId,
+			Zone:   s.zone,
+			Region: s.region,
+			Labels: labelsFromTags(inst.Tags),
+		},
+	}
+	return instInfo, nil
+}
+
+func (s *awsOps) InspectInstanceGroupForInstance(instanceID string) (*cloudops.InstanceGroupInfo, error) {
+	selfInfo, err := s.InspectInstance(instanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	for tag, value := range selfInfo.Labels {
+		// https://docs.aws.amazon.com/autoscaling/ec2/userguide/autoscaling-tagging.html#tag-lifecycle
+		if tag == "aws:autoscaling:groupName" {
+			input := &autoscaling.DescribeAutoScalingGroupsInput{
+				AutoScalingGroupNames: []*string{
+					aws.String(value),
+				},
+			}
 
-func (s *ec2Ops) ApplyTags(volumeID string, labels map[string]string) error {
+			result, err := s.autoscaling.DescribeAutoScalingGroups(input)
+			if err != nil {
+				if aerr, ok := err.(awserr.Error); ok {
+					return nil, aerr
+				}
+				return nil, err
+			}
+
+			if len(result.AutoScalingGroups) != 1 {
+				return nil, fmt.Errorf("DescribeAutoScalingGroups (%v) returned %v groups, expect 1",
+					value, len(result.AutoScalingGroups))
+			}
+
+			group := result.AutoScalingGroups[0]
+			zones := make([]string, 0)
+			for _, z := range group.AvailabilityZones {
+				zones = append(zones, *z)
+			}
+
+			retval := &cloudops.InstanceGroupInfo{
+				CloudResourceInfo: cloudops.CloudResourceInfo{
+					Name:   *group.AutoScalingGroupName,
+					Zone:   s.zone,
+					Region: s.region,
+					Labels: labelsFromTags(group.Tags),
+				},
+				Zones:              zones,
+				AutoscalingEnabled: true,
+				Min:                group.MinSize,
+				Max:                group.MaxSize,
+			}
+
+			return retval, nil
+		}
+	}
+
+	return nil, fmt.Errorf("instance doesn't belong to an instance group")
```

Comment:
this should be a well defined error instead of fmt.Errorf, so that the callers can decide to ignore if they want.

Reply:
done